packages/core/src/services/ai/providers/rules/vercel.test.ts

```diff
@@ -0,0 +1,338 @@
+      },
+    ] as Message[]
+
+    const rules = vercelSdkRules(
+      { rules: [], messages, config },
+      Providers.Anthropic,
+    )
+
+    expect(rules.messages).toEqual([
+      ...(messages.slice(0, 1) as Message[]),
+      {
+        role: 'assistant',
+        content: [{ type: 'text', text: 'Hello! How are you?' }],
+      },
+      {
+        role: 'user',
+        content: [{ type: 'text', text: 'I am good' }],
+      },
+    ])
+  })
+
+  it('put each system message part in a system message and set custom attributes there', () => {
+    messages = [
+      {
+        role: 'system',
+        content: [
+          { type: 'text', text: 'I am a' },
+          { type: 'text', text: 'system message' },
+        ],
+      },
+    ] as Message[]
+
+    const rules = vercelSdkRules(
+      { rules: [], messages, config },
+      Providers.Anthropic,
+    )
+    expect(rules.messages).toEqual([
+      { role: 'system', content: 'I am a' },
+      { role: 'system', content: 'system message' },
+    ])
+  })
+
+  it('already flattened messages', () => {
+    messages = [
+      { role: 'system', content: 'I am a' },
+      { role: 'system', content: 'system message' },
+    ] as unknown as Message[]
+
+    const rules = vercelSdkRules(
+      { rules: [], messages, config },
+      Providers.Anthropic,
+    )
+    expect(rules.messages).toEqual([
+      { role: 'system', content: 'I am a' },
+      { role: 'system', content: 'system message' },
+    ])
+  })
+
+  it('put root system message metadata into text parts', () => {
+    messages = [
+      {
+        role: 'system',
+        cache_control: { type: 'ephemeral' },
+        content: [
+          { type: 'text', text: 'I am a' },
+          { type: 'text', text: 'system message' },
+        ],
+      },
+    ] as Message[]
+
+    const rules = vercelSdkRules(
+      { rules: [], messages, config },
+      Providers.Anthropic,
+    )
+    expect(rules.messages).toEqual([
+      {
+        role: 'system',
+        content: 'I am a',
+        experimental_providerMetadata: {
+          anthropic: {
+            cacheControl: { type: 'ephemeral' },
+          },
+        },
+      },
+      {
+        role: 'system',
+        content: 'system message',
+        experimental_providerMetadata: {
+          anthropic: {
+            cacheControl: { type: 'ephemeral' },
+          },
+        },
+      },
+    ])
+  })
+
+  it('put root user message metadata into text parts', () => {
+    messages = [
+      {
+        role: 'user',
+        cache_control: { type: 'ephemeral' },
+        content: [
+          { type: 'text', text: 'I am a' },
+          { type: 'text', text: 'system message' },
+        ],
+      },
+    ] as Message[]
+
+    const rules = vercelSdkRules(
+      { rules: [], messages, config },
+      Providers.Anthropic,
+    )
+    expect(rules.messages).toEqual([
+      {
+        role: 'user',
+        experimental_providerMetadata: {
```

Comment:
i assume this tranformation only gets applied when the provider is anthropic no?

Reply:
Only when there is metadata. And it ends like this:
``` javascript
{
  experimental_providerMetadata: {
    ['provider_name']: {//....custom tag attributes }
  }
}
```